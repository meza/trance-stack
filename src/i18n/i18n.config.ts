import type { InitOptions } from 'i18next';

export const supportedLngs = ['en'];
export const defaultNS = ['translation'];
export const fallbackLng = 'en';

const baseConfig: InitOptions = {
  supportedLngs: supportedLngs,
  fallbackLng: fallbackLng,
  defaultNS: defaultNS
};

export const getBaseClientConfig = (): InitOptions => ({
  supportedLngs: baseConfig.supportedLngs,
  fallbackLng: baseConfig.fallbackLng,
  defaultNS: baseConfig.defaultNS,
  interpolation: { escapeValue: false },
  react: { useSuspense: false },
  detection: {
    order: ['htmlTag'],
    caches: []
  }
});

export default baseConfig;
