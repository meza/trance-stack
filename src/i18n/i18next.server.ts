import { resolve } from 'node:path';
import { createInstance } from 'i18next';
import Backend from 'i18next-fs-backend';
import { initReactI18next } from 'react-i18next';
import baseConfig from './i18n.config';
import remixI18next from './remix.server';
import type { EntryContext } from '@remix-run/node';
import type { i18n } from 'i18next';

let i18nextInstance: i18n;

const init = async (
  locale: string,
  remixContext: EntryContext
): Promise<i18n> => {
  if (!i18nextInstance) {
    i18nextInstance = createInstance().use(initReactI18next).use(Backend);
    await i18nextInstance.init({
      ...baseConfig,
      lng: locale,
      ns: remixI18next.getRouteNamespaces(remixContext),
      backend: {
        loadPath: resolve('../public/locales/{{lng}}/{{ns}}.json')
      }
    });
  }
  return i18nextInstance;
};

export default init;
