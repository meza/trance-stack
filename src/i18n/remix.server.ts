import { resolve } from 'node:path';
import Backend from 'i18next-fs-backend';
import { RemixI18Next } from 'remix-i18next';
import { supportedLngs, fallbackLng } from './i18n.config';
import type { NewableModule } from 'i18next';

const i18next = new RemixI18Next({
  detection: {
    supportedLanguages: supportedLngs,
    fallbackLanguage: fallbackLng
  },
  i18next: {
    backend: {
      loadPath: resolve('../public/locales/{{lng}}/{{ns}}.json')
    }
  },
  backend: Backend as NewableModule<Backend>
});

export default i18next;
