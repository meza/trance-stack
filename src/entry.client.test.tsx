import { startTransition } from 'react';
import * as Sentry from '@sentry/remix';
import { hydrateRoot } from 'react-dom/client';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { initClientI18n } from '~/i18n';

vi.mock('react');
vi.mock('react-i18next');
vi.mock('~/i18n');
vi.mock('react-dom/client');
vi.mock('i18next', () => ({
  default: 'i18next'
}));
vi.mock('~/components/Mixpanel', () => ({
  Mixpanel: 'Mixpanel'
}));
vi.mock('@remix-run/react');
vi.mock('@sentry/remix');
vi.mock('@sentry/react');

describe('The Client Entrypoint', () => {
  beforeEach(() => {
    vi.resetAllMocks();
    vi.resetModules();
    vi.useRealTimers();
    vi.useFakeTimers();
    vi.mocked(initClientI18n).mockResolvedValue(undefined as never);
    vi.mocked(startTransition).mockImplementation((callback: CallableFunction) => {
      callback();
    });
    vi.mocked(Sentry.init).mockImplementation(() => undefined as never);
    vi.stubGlobal('window', {
      appConfig: {
        sentryDsn: 'sentryDsn',
        version: '0.1.2-dev'
      }
    });
  });

  it('Should initialise Sentry', async () => {
    await import('./entry.client');
    expect(vi.mocked(Sentry.init)).toHaveBeenCalledOnce();
    expect(vi.mocked(Sentry.init).mock.calls[0][0]).toMatchInlineSnapshot(`
      {
        "dsn": "sentryDsn",
        "integrations": [
          BrowserTracing {},
          Replay {},
        ],
        "release": "0.1.2-dev",
        "replaysOnErrorSampleRate": 1,
        "replaysSessionSampleRate": 0.1,
        "tracesSampleRate": 1,
      }
    `);
  });

  it('Should work on Safari', async () => {
    await import('./entry.client');
    await vi.advanceTimersByTimeAsync(1000);

    expect(vi.mocked(hydrateRoot)).toHaveBeenCalledOnce();
  });

  it('Should work on Other Browsers', async () => {
    vi.useRealTimers();
    vi.stubGlobal('requestIdleCallback', (callback: CallableFunction) => {
      callback();
    });
    await import('./entry.client');

    expect(vi.mocked(hydrateRoot)).toHaveBeenCalledOnce();
    const hydrateRootCalls = vi.mocked(hydrateRoot).mock.calls[0];
    expect(hydrateRootCalls[0]).toEqual(document);
    expect(hydrateRootCalls[1]).toMatchSnapshot();
  });
});
