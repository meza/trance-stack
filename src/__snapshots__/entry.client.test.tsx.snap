// Vitest Snapshot v1

exports[`The Client Entrypoint > Should work on Other Browsers 1`] = `
<I18nextProvider
  i18n="i18next"
>
  <Mixpanel />
  <UNDEFINED>
    <RemixBrowser />
  </UNDEFINED>
</I18nextProvider>
`;
